### PR TITLE
DEV: TXI PLCs are not in communication with ARbiter and MR1K3 is in Veto

### PR DESCRIPTION
**Pull request by ghalym**
## Description
All TXI K line PLCs are now connected to the arbiter.
The MR1K3 veto is active. Waiting on other vetos

## Motivation and Context
We need to send beam to TXI next week. We need to bring the TXI K3 plcs into the arbiter and we need to readMR1K3 for veto.

## How Has This Been Tested?
communication tested by manually forcing some values

## Pre-merge checklist
- [ ] Code works interactively
- [ ] Code contains descriptive comments
- [ ] Test suite passes locally
- [ ] Libraries are set to fixed versions and not ``Always Newest``
- [ ] Code committed with ``pre-commit`` (alternatively ``pre-commit run --all-files``)

### CODE REVIEW

**Comment by ZLLentz** [comment]
Hi Maggie, can you please fill out more of the PR description for documentation purposes? At minimum, a link to what we're trying to accomplish?

I can review the PR without this because we work closely together but stuff like this will be helpful when we look back in the future.

File: Arbiter/KFEArbiter/LineIDs.dbg
Comment (ZLLentz) [inline Arbiter/KFEArbiter/LineIDs.dbg:1]: This library's gitignore is out of date, the .dbg files shouldn't be comitted.

File: Arbiter/KFEArbiter/POUs/CurrentBPUpdate.TcPOU [hunk 1]
```diff
@@ -73,7 +73,16 @@ VAR
...
-
+	
+	{attribute 'TcLinkTo' := '
+                .i_StopperInLS:=TIIB[PPS Stoppers 1]^Channel 5^ST1K3 IN;
+                .i_StopperOutLS:=TIIB[PPS Stoppers 1]^Channel 6^ST1K3 OUT;
+                .q_StopperIN_Relay:=TIIB[PMPS_Premp]^Channel 7^ST1K3 RELAY IN;
+                                '}
+    st1k3Watcher : FB_KStopper(
+        PMPS.K_Stopper.ST1K3,
+        'ST1K3');
```
Comment (ZLLentz) [inline Arbiter/KFEArbiter/POUs/CurrentBPUpdate.TcPOU:84]: This ST1K3 stopper is not documented in the PR description

File: Arbiter/KFEArbiter/POUs/CurrentBPUpdate.TcPOU [hunk 2]
```diff
@@ -97,6 +106,35 @@ VAR
         PMPS.K_Stopper.ST4K4,
         'ST4K4');
 
+		
+	 // MR1K3 Veto
+    ////////////////////////
+          {attribute 'TcLinkTo' := 'TIIB[plc-txi-optics]^IO Inputs^MR1K3_Y_ENC'}
+        i_MR1K3_ENC AT %I* : ULINT; // MR1K3 encoder from the txi k optics optics system
```
Comment (ZLLentz) [inline Arbiter/KFEArbiter/POUs/CurrentBPUpdate.TcPOU:113]: Small Q for documentation purposes, is this the horizontal or the vertical encoder? It's not the pitch, right?

File: Arbiter/KFEArbiter/POUs/CurrentBPUpdate.TcPOU [hunk 3]
```diff
@@ -97,6 +106,35 @@ VAR
...
+        cMR1K3_IN_UpperLimit : UDINT := 108500000 + 5000000;
+        cMR1K3_IN_LowerLimit : UDINT := 108500000 - 5000000 ;
+        cMR1K3_OUT_UpperLimit : UDINT := 8900000;
```
Comment (ZLLentz) [inline Arbiter/KFEArbiter/POUs/CurrentBPUpdate.TcPOU:116]: These look correct to me based on the comments in the https://jira.slac.stanford.edu/browse/ECSENG-301 jira if we convert micrometers to nanometers counts :+1:

File: Arbiter/KFEArbiter/POUs/CurrentBPUpdate.TcPOU [hunk 4]
```diff
@@ -97,6 +106,35 @@ VAR
...
+        q_MR1K3_VETO AT %Q* : BOOL; // Signal to accel. MPS that MR1K3 is in
+        fbMR1K3VetoDevice : FB_KVetoDevice(
+            eVetoDeviceIN := K_Stopper.MR1K3_IN,
+            eVetoDeviceOUT := K_Stopper.MR1K3_OUT,
+            sVetoDeviceName := 'MR1K3');
+			
+	// MR2K3 Veto
+    ////////////////////////
+          {attribute 'TcLinkTo' := 'TIIB[plc-txi-optics]^IO Inputs^MR2K3_Y_ENC'}
+        i_MR2K3_ENC AT %I* : ULINT; // MR2K3 encoder from the txi k optics system
```
Comment (ZLLentz) [inline Arbiter/KFEArbiter/POUs/CurrentBPUpdate.TcPOU:127]: Same docs question as before, horizontal or vertical? (hopefully, the same answer)

Follow-up question, not necessarily related to this PR: is there documentation of the geometry here? Does MR2K3 ever block TMO, for example?

File: Arbiter/KFEArbiter/POUs/CurrentBPUpdate.TcPOU [hunk 5]
```diff
@@ -97,6 +106,35 @@ VAR
...
+        cMR2K3_IN_UpperLimit : UDINT := 108500000 + 5000000;
+        cMR2K3_IN_LowerLimit : UDINT := 108500000 - 5000000 ;
+        cMR2K3_OUT_UpperLimit : UDINT := 8900000;
```
Comment (ZLLentz) [inline Arbiter/KFEArbiter/POUs/CurrentBPUpdate.TcPOU:130]: I didn't find any documentation of the limits for MR2K3, but these are the same as MR1K3, so maybe they're correct?

**Comment by ZLLentz** [comment]
In the middle of reviewing, I notice the title says "not" but the PR desc says "now", which one is the correct text?

File: Arbiter/KFEArbiter/POUs/Fast_Faults_Evaluate.TcPOU [hunk 6]
```diff
@@ -7,13 +7,48 @@ FFO    :    FB_FastFault :=(
         i_Desc := 'Fault occurs when the beamline valves on qrixs are not in open state',
         i_TypeCode := 16#1010);
 i_xFastFaultOut_qrix_vac AT %I*:        BOOL;
+
+FFO2   :    FB_FastFault :=(
+        i_Desc := 'Forward Fault from TXI Vacuum.',
+        i_TypeCode := 16#3010);
```
Comment (ZLLentz) [inline Arbiter/KFEArbiter/POUs/Fast_Faults_Evaluate.TcPOU:13]: These type codes are not documented in https://confluence.slac.stanford.edu/display/L2SI/PMPS+Type+Codes

File: Arbiter/KFEArbiter/POUs/Fast_Faults_Evaluate.TcPOU [hunk 7]
```diff
@@ -7,13 +7,48 @@ FFO    :    FB_FastFault :=(
...
+        i_TypeCode := 16#3013);
+{attribute 'TcLinkTo' := 'TIIB[plc-txi-mot]^IO Inputs^FFO'}
+i_xFastFaultOut_txi_mot AT %I*:        BOOL;
+
+
+FFO4    :    FB_FastFault :=(
+        i_Desc := 'Forward Fault from TXI OPTICS.',
+        i_TypeCode := 16#3013);
+{attribute 'TcLinkTo' := 'TIIB[plc-txi-optics]^IO Inputs^FFO'}
+i_xFastFaultOut_txi_optics AT %I*:        BOOL;
```
Comment (ZLLentz) [inline Arbiter/KFEArbiter/POUs/Fast_Faults_Evaluate.TcPOU:28]: In general, FFO2, FFO3, and FFO4 should be documented in the PR description. The PR only mentions the arbitration, not the fault forwarding.

**Review by ZLLentz — COMMENTED** [review]
Your implementation looks good, all my comments are about documentation or miscellaneous things.

The only reason this is "comment" and not "approve" is because I don't know the source for the MR2K3 encoder counts. Did Andy confirm these somewhere?

**Comment by ghalym** [comment]
> Hi Maggie, can you please fill out more of the PR description for documentation purposes? At minimum, a link to context for what we're trying to accomplish?
> 
> I can review the PR without this because we work closely together but stuff like this will be helpful when we look back in the future.

Done

File: Arbiter/KFEArbiter/POUs/CurrentBPUpdate.TcPOU (hunk 1 above)
Comment (ghalym) [reply Arbiter/KFEArbiter/POUs/CurrentBPUpdate.TcPOU:84]: > This ST1K3 stopper is not documented in the PR description

I added it. But I don't have the signal active yet. I thought it is good to have it here for now so it is ready when the signal is active. But for now it is not active.

File: Arbiter/KFEArbiter/POUs/CurrentBPUpdate.TcPOU (hunk 2 above)
Comment (ghalym) [reply Arbiter/KFEArbiter/POUs/CurrentBPUpdate.TcPOU:113]: It is the Y_ENC. Y axis. The name is in the linked input variable

File: Arbiter/KFEArbiter/POUs/CurrentBPUpdate.TcPOU (hunk 4 above)
Comment (ghalym) [reply Arbiter/KFEArbiter/POUs/CurrentBPUpdate.TcPOU:127]: MR2K3 should not be IN when MR1K3 is out. This is fault condition.
Will check the optics to see if it is there.

File: Arbiter/KFEArbiter/POUs/CurrentBPUpdate.TcPOU (hunk 5 above)
Comment (ghalym) [reply Arbiter/KFEArbiter/POUs/CurrentBPUpdate.TcPOU:130]: I am waiting on Andy to give me the values. I have nothing yet. And given that I was given only a short window to deploy changes, I thought I can put these in for now as a place holder.

File: Arbiter/KFEArbiter/POUs/Fast_Faults_Evaluate.TcPOU (hunk 6 above)
Comment (ghalym) [reply Arbiter/KFEArbiter/POUs/Fast_Faults_Evaluate.TcPOU:13]: I didn't get a chance to di it in the small window of working on all the TXI PLCs as well the arbiter. Documentation will follow.

File: Arbiter/KFEArbiter/POUs/CurrentBPUpdate.TcPOU (hunk 5 above)
Comment (ZLLentz) [reply Arbiter/KFEArbiter/POUs/CurrentBPUpdate.TcPOU:130]: Ok! Makes sense, sounds good!

File: Arbiter/KFEArbiter/POUs/CurrentBPUpdate.TcPOU (hunk 2 above)
Comment (ZLLentz) [reply Arbiter/KFEArbiter/POUs/CurrentBPUpdate.TcPOU:113]: d'oh, it's right there and I didn't register in my brain

File: Arbiter/KFEArbiter/POUs/Fast_Faults_Evaluate.TcPOU (hunk 6 above)
Comment (ZLLentz) [reply Arbiter/KFEArbiter/POUs/Fast_Faults_Evaluate.TcPOU:13]: no worries! I understand

**Review by ZLLentz — APPROVED** [review]
The one big confusion I had (MR2K3) is addressed, everything still looks good

**Pull request merged**